### PR TITLE
Include SHA256 checksum in versions v2

### PR DESCRIPTION
**Pull request by kbrock**
## Add `Versions#sha256`

This is a rebase of #713 by @cstrahan - rebase was needed after #610.

- The `sha256` is nullable for now.
- First: populate `sha256` via `rake`.
- Later: add a migration to mark `sha256` as `:null => false`.

### CODE REVIEW

**Comment by indirect** [comment]
Awesome, thanks!

File: app/models/pusher.rb [hunk 1]
```diff
@@ -87,9 +89,12 @@ def find
     # Update the name to reflect a valid case change
     @rubygem.name = name
 
+    sha256 = Digest::SHA2.base64digest(body.string)
```
Comment (arthurnn) [inline app/models/pusher.rb]: why do we store the base64 version? and not the hex as string?

File: app/models/version.rb [hunk 2]
```diff
@@ -256,6 +262,22 @@ def authors_array
...
+  def recalculate_sha256
+    key = "gems/#{full_name}.gem"
+    digest = Digest::SHA2.new
+    dir = Indexer.new.directory
+
+    dir.files.get(key) do |chunk|
+      digest << chunk
+    end
+
+    digest.base64digest
```
Comment (arthurnn) [inline app/models/version.rb]: Also, I kinda dislike this method has a different code-path that the one `pusher` runs... this generation should use the same code path, for the migration and for future generations!

**Comment by arthurnn** [comment]
Hi friends! 
I hate been the guy that reverts this , but we had to. see https://github.com/rubygems/rubygems.org/commit/ec15283

Mainly we revert this because of my 2 comments. and we are going to push other stuff to master, so we dont wanna have deploys blocked.. Lets address those concerns then we can merge this back in.

Thanks all for the hard work!

**Comment by indirect** [comment]
Thanks for the review, and sorry to be the over-eager merger. :grimacing:

File: app/models/pusher.rb (hunk 1 above)
Comment (indirect) [inline app/models/pusher.rb]: Sha265 hashes are _really_ long as hex. I completely agree with this, and strongly prefer to store them as base64. I plan to store them as base64 in the new index as well.

Comment (arthurnn) [inline app/models/pusher.rb]:
```
[arthurnn@ralph rails]$ echo "foo" | shasum -a 256
b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c  -
[arthurnn@ralph rails]$ echo "foo" | shasum -a 256 | base64
YjViYjlkODAxNGEwZjliMWQ2MWUyMWU3OTZkNzhkY2NkZjEzNTJmMjNjZDMyODEyZjQ4NTBiODc4YWU0OTQ0YyAgLQo=
```

Am I missing something?

Comment (indirect) [inline app/models/pusher.rb]: Sha256 hashes are 32 bytes, which means they are 64 characters as hex, or 44 characters as base64. Here's a demo:

```
irb(main):006:0> Digest::SHA256.digest("foo")
=> ",&\xB4kh\xFF\xC6\x8F\xF9\x9BE<\x1D0A4\x13B-pd\x83\xBF\xA0\xF9\x8A^\x88bf\xE7\xAE"
irb(main):007:0> Digest::SHA256.digest("foo").bytes.size
=> 32

irb(main):004:0> Digest::SHA256.hexdigest("foo")
=> "2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae"
irb(main):009:0> Digest::SHA256.hexdigest("foo").bytes.size
=> 64

irb(main):005:0> Digest::SHA256.base64digest("foo")
=> "LCa0a2j/xo/5m0U8HTBBNBNCLXBkg7+g+YpeiGJm564="
irb(main):008:0> Digest::SHA256.base64digest("foo").bytes.size
=> 44
```

Comment (arthurnn) [inline app/models/pusher.rb]: good point.. lets store as base64!

**Comment by arthurnn** [comment]
I just revert-revert this in master and it should be in production soon..
thanks all again!

**Comment by kbrock** [comment]
@arthurnn I'm more than happy to address anything you want. Just say it and I'll add another PR

**Comment by cstrahan** [comment]
@arthurnn What's the state of the checksuming? I see that some gems (e.g. `rails`) have checksums, but I just randomly did a check for the `monads` gem (it was the first thing off the top of my head), and I see that the sha is null:

    λ curl -s rubygems.org/api/v1/versions/monads.json | jq .
    [
      {
        "authors": "Tom Stuart",
        "built_at": "2014-08-02T00:00:00.000Z",
        "description": "",
        "downloads_count": 854,
        "number": "0.0.1",
        "summary": "Simple Ruby implementations of some common monads.",
        "platform": "ruby",
        "ruby_version": ">= 0",
        "prerelease": false,
        "licenses": [
          "MIT"
        ],
        "requirements": [],
        "sha": null
      }
    ]

If they're still being calculated, then that's cool.

----

@kbrock Thanks for wrapping this feature up! :beer:

**Comment by arthurnn** [comment]
@cstrahan right now we are running a backfill to populate the database. as we have a lot of gems and versions that will probably take a while.

**Comment by cstrahan** [comment]
@arthurnn Ah, I just saw your recent commits on master. Awesome - thanks a bunch!

**Pull request merged**